### PR TITLE
azurerm_kubernetes_cluster azurerm_kubernetes_cluster_node_pool - Adding `node_soak_duration_in_minutes` and `drain_timeout_in_minutes` to the node pool `upgrade_settings` block

### PR DESCRIPTION
**Pull request by bitcloud**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

Based on the great PR #24491 from @aristosvo we added the new parameters `node_soak_duration_in_minutes` and `drain_timeout_in_minutes` and updated and fixed the testcases.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [x] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_kubernetes_cluster` `azurerm_kubernetes_cluster_node_pool` - adding `node_soak_duration_in_minutes` and `drain_timeout_in_minutes` to the node pool `upgrade_settings` block
* update `github.com/hashicorp/go-azure-sdk/resource-manager/containerservice` to version `2023-09-02-preview`

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes: #13066
Build on top: #24491

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this PR @bitcloud! Would you mind rebasing this? The version upgrade for AKS just went into main. Once that's done I'll continue with my review.

**Comment by itay-grudev** [comment]
@stephybun I re-based the branch, resolved all the conflicts, ran the tests and tested that drain and soak work correctly with the provider in development mode.

I think it's ready.

File: internal/services/containers/kubernetes_cluster_data_source.go [hunk 1]
```diff
@@ -1465,20 +1465,21 @@ func flattenKubernetesClusterDataSourceMicrosoftDefender(input *managedclusters.
...
 
-	return []interface{}{
-		map[string]interface{}{
-			"max_surge": maxSurge,
-		},
+	if input != nil && input.DrainTimeoutInMinutes != nil {
+		values["node_soak_duration_in_minutes"] = *input.NodeSoakDurationInMinutes
 	}
+
+	return []interface{}{values}
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_data_source.go:1482]: We usually nil check the input first and return early if it's empty, GH won't allow me to use the suggestions feature on this snippet but could we update this to 

```
	if input == nil {
		return []interface{}{}
	}

	values := make(map[string]interface{})
	
	if input.MaxSurge != nil {
		values["max_surge"] = *input.MaxSurge
	}

	if input.DrainTimeoutInMinutes != nil {
		values["drain_timeout_in_minutes"] = *input.DrainTimeoutInMinutes
	}

	if input.DrainTimeoutInMinutes != nil {
		values["node_soak_duration_in_minutes"] = *input.NodeSoakDurationInMinutes
	}

	return []interface{}{values}
```

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 2]
```diff
@@ -1137,6 +1137,14 @@ func upgradeSettingsSchema() *pluginsdk.Schema {
 						Type:     pluginsdk.TypeString,
 						Required: true,
 					},
+					"drain_timeout_in_minutes": {
+						Type:     pluginsdk.TypeInt,
+						Optional: true,
+					},
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource.go:1143]: According to the docs It looks like this has a default value of 30 minutes, so we should set that here as well. Is there also some validation that we could add? e.g. value must be between 0 and 60 minutes?

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 3]
```diff
@@ -1137,6 +1137,14 @@ func upgradeSettingsSchema() *pluginsdk.Schema {
...
+					"node_soak_duration_in_minutes": {
+						Type:     pluginsdk.TypeInt,
+						Optional: true,
+					},
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource.go:1147]: Same here, docs indicate this can be set between 0 and 30 minutes, so we should add that as validation here.

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 4]
```diff
@@ -1152,6 +1160,14 @@ func upgradeSettingsSchema() *pluginsdk.Schema {
...
 					Default:  "10%",
 				},
+				"drain_timeout_in_minutes": {
+					Type:     pluginsdk.TypeInt,
+					Optional: true,
+				},
+				"node_soak_duration_in_minutes": {
+					Type:     pluginsdk.TypeInt,
+					Optional: true,
+				},
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource.go:1170]: Same here, we should apply defaults and validation where applicable

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 5]
```diff
@@ -1223,24 +1247,31 @@ func expandAgentPoolUpgradeSettings(input []interface{}) *agentpools.AgentPoolUp
 	if maxSurgeRaw := v["max_surge"].(string); maxSurgeRaw != "" {
 		setting.MaxSurge = utils.String(maxSurgeRaw)
 	}
+	if drainTimeoutInMinutesRaw, ok := v["drain_timeout_in_minutes"].(int); ok {
+		setting.DrainTimeoutInMinutes = utils.Int64(int64(drainTimeoutInMinutesRaw))
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource.go:1251]: Can we use `pointer.To` here instead 
```suggestion
		setting.DrainTimeoutInMinutes = pointer.To(int64(drainTimeoutInMinutesRaw))
```

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 6]
```diff
@@ -1223,24 +1247,31 @@ func expandAgentPoolUpgradeSettings(input []interface{}) *agentpools.AgentPoolUp
...
+	}
+	if nodeSoakDurationInMinutesRaw, ok := v["node_soak_duration_in_minutes"].(int); ok {
+		setting.NodeSoakDurationInMinutes = utils.Int64(int64(nodeSoakDurationInMinutesRaw))
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource.go:1254]: Same here 
```suggestion
		setting.NodeSoakDurationInMinutes = pointer.To(int64(nodeSoakDurationInMinutesRaw))
```

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 7]
```diff
@@ -1223,24 +1247,31 @@ func expandAgentPoolUpgradeSettings(input []interface{}) *agentpools.AgentPoolUp
...
 
-	return []interface{}{
-		map[string]interface{}{
-			"max_surge": maxSurge,
-		},
+	if input != nil && input.NodeSoakDurationInMinutes != nil {
+		values["node_soak_duration_in_minutes"] = *input.NodeSoakDurationInMinutes
 	}
+
+	return []interface{}{values}
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource.go:1274]: Can we apply the same suggestions to this flatten function as mentioned above

File: internal/services/containers/kubernetes_cluster_node_pool_resource_test.go [hunk 8]
```diff
@@ -565,29 +565,96 @@ func TestAccKubernetesClusterNodePool_upgradeSettings(t *testing.T) {
 
 	data.ResourceTest(t, r, []acceptance.TestStep{
 		{
-			Config: r.upgradeSettingsConfig(data, "2"),
+			Config: r.upgradeSettingsConfig(data, map[string]interface{}{
+				"max_surge":                "2",
+				"drain_timeout_in_minutes": 35,
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource_test.go:570]: Unless there is a good reason to separate testing the update of these properties I would add `node_soak_duration_in_minutes` here and test updating and removing them together. Testing them as individual steps in the same test would lengthen the test duration and drive up the cost/price of it.

File: internal/services/containers/kubernetes_nodepool.go [hunk 9]
```diff
@@ -1829,20 +1837,21 @@ func FlattenDefaultNodePool(input *[]managedclusters.ManagedClusterAgentPoolProf
...
 
-	return []interface{}{
-		map[string]interface{}{
-			"max_surge": maxSurge,
-		},
+	if input != nil && input.NodeSoakDurationInMinutes != nil {
+		values["node_soak_duration_in_minutes"] = *input.NodeSoakDurationInMinutes
 	}
+
+	return []interface{}{values}
```
Comment (stephybun) [inline internal/services/containers/kubernetes_nodepool.go:1854]: Can we apply the same changes to the flatten here as suggested above.

File: internal/services/containers/migration/kubernetes_cluster_node_pool.go [hunk 10]
```diff
@@ -455,6 +455,14 @@ func (k KubernetesClusterNodePoolV0ToV1) Schema() map[string]*pluginsdk.Schema {
...
 						Required: true,
 					},
+					"drain_timeout_in_minutes": {
+						Type:     pluginsdk.TypeInt,
+						Optional: true,
+					},
+					"node_soak_duration_in_minutes": {
+						Type:     pluginsdk.TypeInt,
+						Optional: true,
+					},
```
Comment (stephybun) [inline internal/services/containers/migration/kubernetes_cluster_node_pool.go:465]: Same comments on defaults and validation apply here

File: website/docs/r/kubernetes_cluster.html.markdown [hunk 11]
```diff
@@ -945,6 +945,10 @@ A `http_proxy_config` block supports the following:
 
 A `upgrade_settings` block supports the following:
 
+* `drain_timeout_in_minutes` - (Optional) The amount of time in minutes to wait on eviction of pods and graceful termination per node. This eviction wait time honors waiting on pod disruption budgets. If this time is exceeded, the upgrade fails.
+
+* `node_soak_duration_in_minutes` - (Optional) The amount of time in minutes to wait after draining a node and before reimaging it and moving on to next node. If not specified, the default is 0 minutes.
```
Comment (stephybun) [inline website/docs/r/kubernetes_cluster.html.markdown:950]:
```suggestion
* `node_soak_duration_in_minutes` - (Optional) The amount of time in minutes to wait after draining a node and before reimaging and moving on to next node. Defaults to `0`.
```

File: website/docs/r/kubernetes_cluster_node_pool.html.markdown [hunk 12]
```diff
@@ -308,6 +308,10 @@ A `sysctl_config` block supports the following:
 
 A `upgrade_settings` block supports the following:
 
+* `drain_timeout_in_minutes` - (Optional) The amount of time in minutes to wait on eviction of pods and graceful termination per node. This eviction wait time honors waiting on pod disruption budgets. If this time is exceeded, the upgrade fails.
+
+* `node_soak_duration_in_minutes` - (Optional) The amount of time in minutes to wait after draining a node and before reimaging it and moving on to next node. If not specified, the default is 0 minutes.
```
Comment (stephybun) [inline website/docs/r/kubernetes_cluster_node_pool.html.markdown:313]:
```suggestion
* `node_soak_duration_in_minutes` - (Optional) The amount of time in minutes to wait after draining a node and before reimaging and moving on to next node. Defaults to `0`.
```

File: website/docs/r/kubernetes_cluster.html.markdown [hunk 13]
```diff
@@ -945,6 +945,10 @@ A `http_proxy_config` block supports the following:
 
 A `upgrade_settings` block supports the following:
 
+* `drain_timeout_in_minutes` - (Optional) The amount of time in minutes to wait on eviction of pods and graceful termination per node. This eviction wait time honors waiting on pod disruption budgets. If this time is exceeded, the upgrade fails.
```
Comment (stephybun) [inline website/docs/r/kubernetes_cluster.html.markdown:948]: > This eviction wait time honors waiting on pod disruption budgets

I'm unsure if this is correct, in the context of deleting the cluster we actually ignore PDBs

File: website/docs/d/kubernetes_cluster.html.markdown [hunk 14]
```diff
@@ -182,6 +182,11 @@ An `azure_active_directory_role_based_access_control` block exports the followin
 
 A `upgrade_settings` block exports the following:
 
+
+* `drain_timeout_in_minutes` - The amount of time in minutes to wait on eviction of pods and graceful termination per node. This eviction wait time honors waiting on pod disruption budgets. If this time is exceeded, the upgrade fails.
+
+* `node_soak_duration_in_minutes` - The amount of time in minutes to wait after draining a node and before reimaging it and moving on to next node. If not specified, the default is 0 minutes.
+
```
Comment (stephybun) [inline website/docs/d/kubernetes_cluster.html.markdown:189]: Looks like there is an redundant space here and we also don't need to specify any defaults or possible values for properties in the data source docs
```suggestion
* `drain_timeout_in_minutes` - The amount of time in minutes to wait on eviction of pods and graceful termination per node. This eviction wait time honors waiting on pod disruption budgets. If this time is exceeded, the upgrade fails.

* `node_soak_duration_in_minutes` - The amount of time in minutes to wait after draining a node and before reimaging it and moving on to next node.

```

File: website/docs/d/kubernetes_cluster_node_pool.html.markdown [hunk 15]
```diff
@@ -94,6 +94,10 @@ In addition to the Arguments listed above - the following Attributes are exporte
 
 A `upgrade_settings` block exports the following:
 
+* `drain_timeout_in_minutes` - The amount of time in minutes to wait on eviction of pods and graceful termination per node. This eviction wait time honors waiting on pod disruption budgets. If this time is exceeded, the upgrade fails.
+
+* `node_soak_duration_in_minutes` - The amount of time in minutes to wait after draining a node and before reimaging it and moving on to next node. If not specified, the default is 0 minutes.
```
Comment (stephybun) [inline website/docs/d/kubernetes_cluster_node_pool.html.markdown:99]:
```suggestion
* `node_soak_duration_in_minutes` - The amount of time in minutes to wait after draining a node and before reimaging it and moving on to next node.
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @bitcloud.

I made a first pass through this. Could you resolve the review comments? Once that's done we can run the tests and see if we're good to go or require more changes.

File: website/docs/r/kubernetes_cluster.html.markdown (hunk 13 above)
Comment (itay-grudev) [reply website/docs/r/kubernetes_cluster.html.markdown:948]: The text is more about upgrades. I would intuitively expect that it is ignored for cluster deletion.

Comment (stephybun) [reply website/docs/r/kubernetes_cluster.html.markdown:948]: I haven't done much with PDBs so it wasn't implicit for me. I think to avoid any potential confusion it'd be good to 
specify that e.g. 
```suggestion
* `drain_timeout_in_minutes` - (Optional) The amount of time in minutes to wait on eviction of pods and graceful termination per node. This eviction wait time honors pod disruption budgets for upgrades. If this time is exceeded, the upgrade fails.
```

**Comment by itay-grudev** [comment]
@stephybun I added all of the requested changes.

I am unsure about one of your documentation comments. Could you give me more guidance if you'd like it changed.

I trimmed the tests, but I left the corner cases with 0 values.

**Comment by stephybun** [comment]
@itay-grudev 

>I am unsure about one of your documentation comments. Could you give me more guidance if you'd like it changed.

Sure, could you specify which?

**Comment by itay-grudev** [comment]
> @itay-grudev
> 
> > I am unsure about one of your documentation comments. Could you give me more guidance if you'd like it changed.
> 
> Sure, could you specify which?

Never mind. You already replied to my comment 20min ago. I just applied your proposal.

Are you ok with the tests like this? I combined two of the tests in one, but I feel like the 0s and empty configuration should remain.

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 16]
```diff
@@ -1137,6 +1137,16 @@ func upgradeSettingsSchema() *pluginsdk.Schema {
...
 					},
+					"drain_timeout_in_minutes": {
+						Type:     pluginsdk.TypeInt,
+						Optional: true,
+						Default:  30,
+					},
+					"node_soak_duration_in_minutes": {
+						Type:     pluginsdk.TypeInt,
+						Optional: true,
+						Default:  0,
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource.go:1148]: Integers that don't have a `Default` specified in the schema have a default of `0` anyway so this isn't necessary, what we can and should add here is the validation I mentioned earlier 
```suggestion
						ValidateFunc: validation.IntBetween(0, 30),
```
Apologies that wasn't clear earlier.

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 17]
```diff
@@ -1152,6 +1162,16 @@ func upgradeSettingsSchema() *pluginsdk.Schema {
...
 				},
+				"drain_timeout_in_minutes": {
+					Type:     pluginsdk.TypeInt,
+					Optional: true,
+					Default:  30,
+				},
+				"node_soak_duration_in_minutes": {
+					Type:     pluginsdk.TypeInt,
+					Optional: true,
+					Default:  0,
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource.go:1173]:
```suggestion
					ValidateFunc: validation.IntBetween(0, 30),
```

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 18]
```diff
@@ -1167,6 +1187,16 @@ func upgradeSettingsForDataSourceSchema() *pluginsdk.Schema {
...
+				"drain_timeout_in_minutes": {
+					Type:     pluginsdk.TypeInt,
+					Optional: true,
+					Default:  "30",
+				},
+				"node_soak_duration_in_minutes": {
+					Type:     pluginsdk.TypeInt,
+					Optional: true,
+					Default:  0,
+				},
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource.go:1199]: This is the schema for the data source so these should only be `Computed`

```suggestion
				"drain_timeout_in_minutes": {
					Type:     pluginsdk.TypeInt,
					Computed: true,
				},
				"node_soak_duration_in_minutes": {
					Type:     pluginsdk.TypeInt,
					Computed: true,
				},
```

File: internal/services/containers/migration/kubernetes_cluster_node_pool.go [hunk 19]
```diff
@@ -455,6 +455,16 @@ func (k KubernetesClusterNodePoolV0ToV1) Schema() map[string]*pluginsdk.Schema {
...
+					"drain_timeout_in_minutes": {
+						Type:     pluginsdk.TypeInt,
+						Optional: true,
+						Default:  30,
+					},
+					"node_soak_duration_in_minutes": {
+						Type:     pluginsdk.TypeInt,
+						Optional: true,
+						Default:  0,
+					},
```
Comment (stephybun) [inline internal/services/containers/migration/kubernetes_cluster_node_pool.go:467]: Overlooked this in the last review, but state migrations refer to the schema at a specific point in time, so these need to be removed 
```suggestion
```

File: internal/services/containers/kubernetes_cluster_resource_upgrade_test.go [hunk 20]
```diff
@@ -279,29 +279,64 @@ func TestAccKubernetesCluster_upgradeSettings(t *testing.T) {
 
 	data.ResourceTest(t, r, []acceptance.TestStep{
 		{
-			Config: r.upgradeSettingsConfig(data, "2"),
+			Config: r.upgradeSettingsConfig(data, map[string]interface{}{
+				"max_surge":                "2",
+				"drain_timeout_in_minutes": 35,
+			}),
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_resource_upgrade_test.go:285]: These test steps could still be condensed further, I've added what the test steps should look like below:

```
data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.upgradeSettingsConfig(data, map[string]interface{}{
				"max_surge":                    "2%",
				"drain_timeout_in_minutes":      35,
				"node_soak_duration_in_minutes": 18,
			}),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.#").HasValue("1"),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.0.max_surge").HasValue("2"),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.0.drain_timeout_in_minutes").HasValue("40"),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.0.node_soak_duration_in_minutes").HasValue("18"),
			),
		},
		data.ImportStep(),
		{
			Config: r.upgradeSettingsConfig(data, map[string]interface{}{
				"max_surge":                    "10%",
				"drain_timeout_in_minutes":      0,
				"node_soak_duration_in_minutes": 0,
			}),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.#").HasValue("1"),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.0.max_surge").HasValue("10%"),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.0.drain_timeout_in_minutes").HasValue("0"),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.0.node_soak_duration_in_minutes").HasValue("0"),
			),
		},
		data.ImportStep(),
		{
			Config: r.upgradeSettingsConfig(data, map[string]interface{}{
				"max_surge": "10%",
			}),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.#").HasValue("1"),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.0.max_surge").DoesNotExist(),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.0.drain_timeout_in_minutes").DoesNotExist(),
				check.That(data.ResourceName).Key("default_node_pool.0.upgrade_settings.0.node_soak_duration_in_minutes").DoesNotExist(),
			),
		},
		data.ImportStep(),
```

This will:
1. Test the create path for `drain_timeout_in_minutes` and `node_soak_duration_in_minutes`
2. Test the update path for `drain_timeout_in_minutes` and `node_soak_duration_in_minutes`
3. Test the update path for `drain_timeout_in_minutes` and `node_soak_duration_in_minutes` by removing it from the config

As a note, `max_surge` _needs_ to be specified at the moment since the API changed the default (this will be defaulted in the provider in 4.0). We get a plan diff if it isn't specified.

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @bitcloud @itay-grudev

Apologies if some of my review comments were unclear. I've tried to clarify and be more specific in my suggestions this time around. I also missed that the state migration schema was updated in my initial review, this needs to be removed. Please let me know if you have any further questions. Once these comments are resolved I'll run the tests and we'll see if any further changes are required.

**Comment by itay-grudev** [comment]
@stephybun Done. I'm sorry I took so long. I was waiting for the tests to finish.

**Comment by itay-grudev** [comment]
@stephybun

```
:-) 
```

Thanks for the help.

**Comment by itay-grudev** [comment]
@stephybun Can we do anything to help you and accelerate this to `main`?

**Comment by stephybun** [comment]
@itay-grudev we have some test failures due to the way the `max_surge` property is handled in AKS at the moment which will change in the next upcoming major release of the provider. This property may also require some special handling up until 4.0 as well. Would it be alright for you if I pushed some changes to your branch to get this into a state where we can merge it?

**Comment by itay-grudev** [comment]
@stephybun Yhea, I don't mind of course. Ping me if I can assist in any way.

**Comment by stephybun** [comment]
Apologies for needing to intervene here @bitcloud / @itay-grudev.

The AKS API can be tricky at times, and while looking into this I found that the `drain_timeout_in_minutes` property cannot be set to `0` or removed/reset back to it's default after it's been specified. This requires a `CustomizeDiff` to force recreate the resource if a user attempts to unset this property since it otherwise results in a permanent plan diff. This also meant the tests could be greatly simplified. 

I unfortunately did not have the necessary permissions to push those updates to your branch, and given the large number of changes made I've opened a separate PR that supersedes this one. I did my best to maintain the authorship of the original work in https://github.com/hashicorp/terraform-provider-azurerm/pull/26137. I'm going to close this PR now in favour of #26137. The tests are running for this and I'll have to defer review of this to another member of the team but please subscribe to #26137 for updates. 

Thanks again for all your effort on this!
